Solution level header definition

RandomEngy opened the pull request:
This supercedes the other pull request.

* Added ability to add a solution-level license header definition file. It is kept at SolutionName.sln.licenseheader .
* Refactored command names to be more clear about what they do.
* Changed "Add header to all projects" to "Add header to all files in solution" to more accurately describe the effects.
* Updated .gitignore to ignore .vs folder
* Added .editorconfig to automatically let editors enforce some of the coding conventions the project uses

FlorianDecker commented:
Thanks for your contribution! We discussed you changes and have some general comments:

- Please keep the long form "LicenseHeader" instead of "Header" in the code. There is a trade-off between readability and clarity (with possible redundancy) and we would like to keep the clarity with the long form of the name.
- The rename to "Add header to all projects" is much more accurate, nice change!
- The "Add New License Header Definition File" button on the solution layer should be greyed out instead of hiding it if the file already exists. I know that we currently hide the Add/Remove LicenseHeader Buttons on items, I documented our argumentation of the behaviour [here](https://github.com/rubicon-oss/LicenseHeaderManager/wiki/When-should-a-Context-Menu-Button-be-greyed-out-and-when-should-it-be-hidden%3F), hopefully it makes sense.
- Please rename the new "Add" button to "Add New Solution License Header Definition", to highlight that this .licenseheader file behaves different than the other .licenseheader files.
- As the solution wide licenseheader file does not have to exist inside the solution, we should add an "Remove Solution License Header Definition File" and "Open Solution License Header Definition File" below the "Add New LicenseHeader Definition File". This way it is possible to 100% control the .licenseheader file from the IDE. The "Remove" Button should be greyed out if no solution .licenseheader file exists.
- We dont think that the .licenseheader should be added to the solution items folder. Adding the file automatically into the solution could suggest to the user, that the .licenseheader file behaves hierarchically like the other .licenseheader files. With this change, we practically add a "special" .licenseheader file which behaves differently than the other .licenseheader files. It has the additional constraint that it has to be named exactly like the solution (for the other .licenseheader files only their location in the file hierarchie matters). Placing it in a solution folder could suggest that the .licenseheader file is applied for the solution folder only (so practically the behaviour i describe in #58, which still could be added as a sepearate feature). Leaving it outside the solution makes it seem more like a solution wide setting file. Also, if a user deletes the file from the solution items, it is not deleted from the disk and is still applied. The user can add it manually into their solution folder if they want to, with this action they hopefully are aware that this file behaves differently than the other.

As a follow up convenience feature, the new "Remove" button could also check if the file was added to solution and if yes, delete it from the solution too.

RandomEngy commented:
I did the rebase onto the F5 fix. All your thoughts about not adding to the solution items folder make sense; it was a little weird to be able to remove it but have it still function. I think your idea of controlling it entirely from context menu options is good. Can also do the grey out instead of disable change.

Though I'm a bit surprised that you think saying "Header" instead of "LicenseHeader" is less clear. I think in the context of the project there's only one thing that can be referring to. And some of those names are pretty long already; I felt it was more important to clearly call out what scope the methods were applying to. But, I can still add the word back in if you really want it.

I'll see if I can get some time to work on the revisions today or tomorrow.

RandomEngy commented:
Okay, I added a couple new commands (with new icons, added to the .pdn), and changed to be enable/disable and not add to Solution Items.

Let me know if you really want the "License" term back in all the command names. I find the intent clear without it but I can always add it back.

FlorianDecker commented:
Thanks for the changes! 

I reviewed your renames again and yes, please add the "Licence" Term back.

There is a minimum of two different meainings for "Header" inside the solution. The second beside LicenseHeader is Document header (where we just word with a commented header from the top of a file, we dont know yet if this is a licenseheader).

If i am not mistaken, we seem to use only "Header" in the method name if the type of header (DocumentHeader/LicenseHeader) is clear from the name of the class (its possible that we are not 100% consistent). While your change would shorten class/method names, it would not improve consistency and clarity.

FlorianDecker commented on LicenseHeaderManager/Utils/LinkedFileFilter.cs:
```diff
@@ -14,6 +14,7 @@
 
 using System.Collections.Generic;
 using EnvDTE;
+using EnvDTE80;
```
unused using. But as you seem to use that consistently, am i missing something?

RandomEngy replied:
No, I think it just got added automatically as I was working on the PR. I'll clean these up.

FlorianDecker commented on LicenseHeaderManager/Utils/LinkedFileHandler.cs:
```diff
@@ -15,6 +15,7 @@
 using System.Collections.Generic;
 using System.Linq;
 using EnvDTE;
+using EnvDTE80;
```
Unused using

FlorianDecker commented on LicenseHeaderManager/Headers/LicenseHeaderFinder.cs:
```diff
@@ -17,6 +17,7 @@
 using System.IO;
 using System.Linq;
 using EnvDTE;
+using EnvDTE80;
```
unused using

FlorianDecker commented on LicenseHeaderManager/PackageCommands/AddNewSolutionHeaderDefinitionFileCommand.cs:
```diff
@@ -0,0 +1,52 @@
+      string solutionHeaderDefinitionFilePath = LicenseHeader.GetHeaderDefinitionFilePathForSolution(solution);
+
+      // Add file
+      string defaultLicenseHeaderFileText = this.defaultHeaderDefinitionFunc();
```
Please rename "_defaultHeaderDefinitionFunc" and remove the "this."

FlorianDecker commented on LicenseHeaderManager/PackageCommands/AddHeaderToAllFilesInProjectCommand.cs:
```diff
@@ -14,17 +14,19 @@
-    public AddLicenseHeaderToAllFilesCommand(LicenseHeaderReplacer licenseReplacer)
+    public AddHeaderToAllFilesInProjectCommand(LicenseHeaderReplacer licenseReplacer)
     {
       this.licenseReplacer = licenseReplacer;
```
Thats a mistake by me, the private field should be named "_licenseReplacer" and the "this." should be removed. Just as information, as you used this naming scheme later on. If you have time you could clean it up, but i can do that myself later on. Just use the correct notation for the newly added code.

FlorianDecker commented on LicenseHeaderManager/PackageCommands/AddNewSolutionHeaderDefinitionFileCommand.cs:
```diff
@@ -0,0 +1,52 @@
+
+    private AddNewSolutionHeaderDefinitionFileCommand(Func<string> defaultHeaderDefinitionFunc)
+    {
+      this.defaultHeaderDefinitionFunc = defaultHeaderDefinitionFunc;
```
Please rename "_defaultHeaderDefinitionFunc" and remove the "this."

FlorianDecker commented on LicenseHeaderManager/PackageCommands/RemoveHeaderFromAllFilesInProjectCommand.cs:
```diff
@@ -17,11 +17,11 @@
-    public RemoveLicenseHeaderFromAllFilesCommand(LicenseHeaderReplacer licenseReplacer)
+    public RemoveHeaderFromAllFilesInProjectCommand(LicenseHeaderReplacer licenseReplacer)
     {
       this.licenseReplacer = licenseReplacer;
```
Again, mistake by me... i am going to clean this up myself after the pull request

FlorianDecker commented on LicenseHeaderManager/PackageCommands/RemoveHeaderFromAllFilesInSolutionCommand.cs:
```diff
@@ -20,14 +20,14 @@
+    public RemoveHeaderFromAllFilesInSolutionCommand(IVsStatusbar statusBar, LicenseHeaderReplacer licenseReplacer)
     {
       this._statusBar = statusBar;
       this._licenseReplacer = licenseReplacer;
```
Again, mistake by me... i am going to clean this up myself after the pull request

FlorianDecker commented on .editorconfig:
```diff
@@ -1,3 +1,7 @@
-indent_size = 2
\ No newline at end of file
+indent_size = 2
+dotnet_style_qualification_for_field = false
```
This seems to have happened mainly because there was some inconsistency in the existing code. See further comments.

RandomEngy replied:
Yeah Resharper has its own rules that was overriding the Visual Studio ones. I wasn't sure if you used Resharper so I was just trying to manually correct where it was adding them, but I guess I missed a few places. If you like I can check in a .sln.DotSettings file that will apply the style for Resharper as well.

RandomEngy commented:
Okay did the rename and cleanup requested. I also snuck in one more change: I added "©" to the default list of things it searches for to find a copyright header.

FlorianDecker commented on LicenseHeaderManager/PackageCommands/AddLicenseHeaderToAllFilesInSolutionCommand.cs:
```diff
@@ -0,0 +1,100 @@
+          ? Resources.Question_SomeProjectsMissingAddNewLicenseHeaderDefinitionForSolution
+          : Resources.Question_AddNewLicenseHeaderDefinitionForSolution;
+
+        if (MessageBoxHelper.DoYouWant(question))
```
It seems like you changed the behaviour quite a lot here. We would like to add the new solution header file as additional/optional feature, which does not completely replace the existing "Definition File per Project" Pattern. We can discuss where the new solution header would make sense.
 
1. If only one Project exists and it has no Definition file -> Offer to add to project (should not be changed, if we have only one project we dont need a solution wide definition file)
    1. If the offer is declined, nothing happens
    2. If the offer is approved, add file to project and offer to stop update process for configuring
        1. If this offer is declined, update license headers in project
2. If multiple Projects exists and none of them have a Definition file -> Offer to add new ones for every one of them (here it would maybe make sense to offer the solution wide License header file? I will check if my colleague has concerns regarding that change and come back to you)
    1. If the offer is declined, nothing happens
    2. If the offer is approved, add file to project and offer to stop update process for configuring
        1. If this offer is declined, update license headers in projects
3. If multiple Projects exists and some of them have a Definition file -> Offer to add existing ones for the missing projects (should not be changed, if some of them are missing a definition file it either wants to skip these projects for updating or currently uses the "definition file per project" pattern. If they want to change to solution wide, they have to do that manually.)
    1. If the offer is declined, **update the license headers for all projects**
    2. If the offer is approved, add existing file to project and offer to stop update process for configuring
        1. If this offer is declined, update license headers in all projects

Note: It seems like we currently only check for the definition file in the root of the project and dont discover them if they are inside a folder in a project. Not sure if that is a bug or a feature. This means that, to keep the current update behaviour in tact, 3. i) should definitely keep updating all projects in the solution, independent if we discovered a definition file there.

If we define the current behaviour (which i described in the "Note:") as a bug, we will fix that in a following branch. So for this pull request, keep the current behaviour, except maybe change "2." to the new solution definition file. I should have an answer from my colleague in the next hour or so.

RandomEngy replied:
It sounds like we are in agreement for scenarios 1 and 2.

For point 3, my thought is that users in most cases are going to want to have the same header definitions across their solution, with specific exceptions for specific projects and folders. At least with the dozens of solutions I've seen I've never had somebody want to have different headers for different projects. It's great that it's supported, but I don't think it's the common use case.

It seems to me that the "add existing header file definition" is a feature that was designed to make sharing the same header definition file across the solution easier. But it's really not the ideal way to do it because you need to manually add the link to new projects and you need to keep the link up to date as you move projects around. It doesn't make sense to me to keep code around to continue propping up and promoting this stopgap measure when we have something better around. Adding the solution level definition is the least invasive (as you only ever need to add one file) and it future proofs you for additional projects. Prompting in this manner also might make users aware of the better alternative.

Anyway I've made the renames requested. I think the new workflow for handling missing header definitions is a lot cleaner and pushes users toward the best way of solving their problem, but let me know if you'd really want to preserve the "add existing header" workflow.

FlorianDecker replied:
While i agree with the general sentiment, i dont think that adding the solution definition file in this dialog is a clean solution. Would we add that file, the user would now use two different systems in his solution: the solution definition file and an unknow amount of other definition files. This could lead to confusion if he only maintains one of these systems and forgets to change the other one. 
I think that step should be actively taken and not happen because the user clicked "yes" in a dialog. Especially because the user should probably clean up the old system afterwards.

To make it clean we could offer some kind of migration scenario and offer to consilidate the existing definition files into a single solution definition file. This would have to be limited to a small scope, as it could get complex to have the same behaviour for all possible use cases. We could think about this in a follow up task.

Also, while the new solution definition file is much better for people having simple license requirements, the current system is more flexible for multiple use cases. We shouldnt make the usage harder for people who maybe need that kind of complexity.

We could inform the user of the new possibility in the dialog though. Or maybe offer both solutions in the dialog with customized buttons (altough i am not sure if that is possible with those dialogs).

RandomEngy replied:
I've changed to try to prompt for an existing file when some projects have it, but others don't. But note that AddExistingLicenseHeaderDefinitionFileToProjectCommand.OpenFileDialogForExistingFile() does not seem to work for VS2017 (and is broken in your current release). That will be a separate bug for you to fix. I also added a check to filter out the "miscellaneous items" project when determining if any projects had a definition file because that showed up in the list even when there was nothing on the solution.

RandomEngy commented:
Hmm, not sure where my reply went. Anyway I made the requested change, though AddExistingLicenseHeaderDefinitionFileToProjectCommand.OpenFileDialogForExistingFile() is broken in the current version and is still broken after my changes. Opened https://github.com/rubicon-oss/LicenseHeaderManager/issues/91 to track.

FlorianDecker commented on LicenseHeaderManager/LicenseHeadersPackage.cs:
```diff
@@ -507,50 +536,22 @@ private void AddLicenseHeadersToAllFilesCallback (object sender, EventArgs e)
-
-        //If there is a licenseheader in the Solution 
-        if(projects.Any(projectInSolution => LicenseHeaderFinder.GetHeader(projectInSolution) != null))
+        // No license header found...
```
Same argumentation as on the solution level. If another project contains a .licenseheader file, offer to link existing .licenseheader file (seems to have worked in contrary to the same action on solution level). If no project contains a .licenseheader file, we can offer the solution wide definition file.

FlorianDecker commented on LicenseHeaderManager/LicenseHeadersPackage.cs:
```diff
@@ -341,7 +370,7 @@ private void AfterLinkedCommandExecuted (string guid, int id, object customIn, o
     {
       _isCalledByLinkedCommand = true;
-      _addLicenseHeaderCommand.Invoke (false);
+      this._addHeaderCommand.Invoke (false);
```
remove this.

FlorianDecker commented on LicenseHeaderManager/LicenseHeadersPackage.cs:
```diff
@@ -232,8 +246,8 @@ private void QueryProjectItemCommandStatus (object sender, EventArgs e)
-      _addLicenseHeaderToProjectItemCommand.Visible = visible;
-      _removeLicenseHeaderFromProjectItemCommand.Visible = visible;
+      this._addHeaderToProjectItemCommand.Visible = visible;
+      this._removeHeaderFromProjectItemCommand.Visible = visible;
```
remove this.

FlorianDecker commented on LicenseHeaderManager/LicenseHeadersPackage.cs:
```diff
@@ -214,8 +228,8 @@ private void QueryEditCommandStatus (object sender, EventArgs e)
 
-      _addLicenseHeaderCommand.Visible = visible;
-      _removeLicenseHeaderCommand.Visible = visible;
+      this._addHeaderCommand.Visible = visible;
```
remove this.

RandomEngy commented:
Addressed the last comments.

FlorianDecker commented:
Nice, looks good! Merged.

Thanks for contributing :) We are going to release it as official version as soon as it is possible to update Visual Studio Extensions on the new Visual Studio Marketplace (which, according to microsoft, should be possible in the next weeks).

RandomEngy commented:
Horaay! You could still publish the update to the old site for now; they will automatically move everything over.

FlorianDecker commented:
I know that its possible, its just that updating via the old gallery is quite a hassle and has to be done by a colleague. On the new marketplace i should be possible to upload it myself and it should be a more pleasent experience :)

RandomEngy commented:
Fair enough. Thanks!

RandomEngy commented:
Looks like you can update extensions on the new Marketplace now!

FlorianDecker commented:
Thanks for the bump! Extension is updated on the marketplace. Thank you for your contribution :)
